CodeFlareSDK_Design_Doc.md

```diff
@@ -0,0 +1,113 @@
+
+* We do not want to re-implement any existing SDK’s or clients for Ray, MCAD, or any other service.
+
+## Architecture and Design
+
+The CodeFlare SDK is a python package that allows a user to programmatically define, create, monitor, and shutdown framework clusters (RayClusters for now) as well as define, submit, monitor and cancel distributed training jobs for machine learning models on an authenticated Kubernetes cluster that has the CodeFlare Stack installed.
+
+In order to achieve this we need the capacity to:
+
+* Generate valid appwrapper yaml files based on user provided parameters
+* Get, list, watch, create, update, patch, and delete appwrapper custom resources on a kubernetes cluster
+* Get, list, watch, create, update, patch, and delete RayCluster custom resources on a kubernetes cluster.
+* Expose a secure route to the  Ray Dashboard endpoint.
+* Define, submit, monitor and cancel Jobs submitted via Torchx. TorchX jobs must support both Ray and MCAD-Kubernetes scheduler backends.
+* Provide means of authenticating to a Kubernetes cluster
+
+![](/docs/images/sdk-diagram.png)
+
+### Framework Clusters:
+
+In order to create these framework clusters, we will start with a template AppWrapper yaml file with reasonable defaults that will generate a valid RayCluster via MCAD.
+
+Users can customize their AppWrapper by passing their desired parameters to `ClusterConfig()` and applying that configuration when initializing a `Cluster()` object.  When a `Cluster()` is initialized, it will update the AppWrapper template with the user’s specified requirements, and save it to the current working directory.
+
+Our aim is to simplify the process of generating valid AppWrappers for RayClusters, so we will strive to find the appropriate balance between ease of use and exposing all possible AppWrapper parameters. And we will find this balance through user feedback.
+
+With a valid AppWrapper, we will use the Kubernetes python client to apply the AppWrapper to our Kubernetes cluster via a call to `cluster.up()`
+
+We will also use the Kubernetes python client to get information about both the RayCluster and AppWrapper custom resources to monitor the status of our Framework Cluster.
+
+The RayCluster deployed on your kubernetes cluster can be interacted with in two ways: Either through an interactive session via `ray.init()` or through the submission of batch jobs.
+
+Finally we will use the Kubernetes python client to delete the AppWrapper via `Cluster.down()`
+
+### Training Jobs:
+
+For the submission of Jobs we will rely on the [TorchX](https://pytorch.org/torchx/latest/) job launcher to handle orchestrating the distribution of our model training jobs across the available resources on our cluster. We will support two distributed backend schedulers: Ray and Kuberentes-MCAD. Torchx is designed to be used primarily as a CLI, so we will wrap a limited subset of its functionality into our SDK so that it can be used as part of a python script.
```

Comment:
```suggestion
For the submission of Jobs we will rely on the [TorchX](https://pytorch.org/torchx/latest/) job launcher to handle orchestrating the distribution of our model training jobs across the available resources on our cluster. We will support two distributed backend schedulers: Ray and Kuberentes-MCAD. TorchX is designed to be used primarily as a CLI, so we will wrap a limited subset of its functionality into our SDK so that it can be used as part of a python script.
```